Sources/MongoSwift/ReadPreference.swift

```diff
@@ -0,0 +1,167 @@
+        }
+
+        return mode
+    }
+
+    /// The tags of this `ReadPreference`
+    public var tagSets: [Document] {
+        guard let bson = mongoc_read_prefs_get_tags(self._readPreference) else {
+            return []
+        }
+
+        let wrapped = Document(fromPointer: bson)
+
+        return wrapped.values as! [Document]
+    }
+
+    /// The maxStalenessSeconds of this `ReadPreference`
+    public var maxStalenessSeconds: Int64? {
+        let maxStalenessSeconds = mongoc_read_prefs_get_max_staleness_seconds(self._readPreference)
+
+        if maxStalenessSeconds == MONGOC_NO_MAX_STALENESS {
+            return nil
+        }
+
+        return maxStalenessSeconds
+    }
+
+    /**
+     * Initializes a `ReadPreference` from a `Mode`.
+     *
+     * - Parameters:
+     *   - mode: a `Mode`
+     *
+     * - Returns: a new `ReadPreference`
+     */
+    public init(_ mode: Mode) {
+        self._readPreference = mongoc_read_prefs_new(mode.readMode)
+    }
+
+    /**
+     * Initializes a `ReadPreference`.
+     *
+     * It is an error to combine a non-empty tagSets value with a primary mode.
+     * If specified, maxStalenessSeconds must be at least 90.
+     *
+     * - Parameters:
+     *   - mode: a `Mode`
+     *   - tagSets: an optional `[Document]`
+     *   - maxStalenessSeconds: an optional Int64
+     *
+     * - Returns: a new `ReadPreference`
+     */
+    public init(_ mode: Mode, tagSets: [Document]? = nil, maxStalenessSeconds: Int64? = nil) throws {
+        self._readPreference = mongoc_read_prefs_new(mode.readMode)
+
+        if let tagSets = tagSets {
+            guard mode != .primary || tagSets.isEmpty else {
+                throw MongoError.readPreferenceError(message: "tagSets may not be used with primary mode")
+            }
+
+            let encoder = BsonEncoder()
+            let tags = try encoder.encode(Document(tagSets))
+            mongoc_read_prefs_set_tags(self._readPreference, tags.data)
+        }
+
+        if let maxStalenessSeconds = maxStalenessSeconds {
+            guard maxStalenessSeconds >= MONGOC_SMALLEST_MAX_STALENESS_SECONDS else {
+                throw MongoError.readPreferenceError(message: "Expected maxStalenessSeconds to be >= \(MONGOC_SMALLEST_MAX_STALENESS_SECONDS), \(maxStalenessSeconds) given")
+            }
+
+            mongoc_read_prefs_set_max_staleness_seconds(self._readPreference, maxStalenessSeconds);
+        }
+    }
+
+    /// Initializes a new `ReadPreference` by copying an existing `ReadPreference`.
+    public init(from readPreference: ReadPreference) {
+        self._readPreference = mongoc_read_prefs_copy(readPreference._readPreference)
+    }
+
+    /// Initializes a new `ReadPreference` by copying a `mongoc_read_prefs_t`.
+    /// The caller is responsible for freeing the original `mongoc_read_prefs_t`.
+    internal init(from readPreference: OpaquePointer?) {
+        self._readPreference = mongoc_read_prefs_copy(readPreference)
```

Comment:
I don't yet have a way to test this, but I expect it will be tested when we support ReadPreference accessors on the core classes (e.g. MongoClient).